pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -309,25 +343,21 @@ func (s *Scheduler) add(ctx context.Context, pod *corev1.Pod) error {
 	// Create new node
 	var errs error
 	for _, nodeClaimTemplate := range s.nodeClaimTemplates {
-		instanceTypes := nodeClaimTemplate.InstanceTypeOptions
-		// if limits have been applied to the nodepool, ensure we filter instance types to avoid violating those limits
-		if remaining, ok := s.remainingResources[nodeClaimTemplate.NodePoolName]; ok {
-			instanceTypes = filterByRemainingResources(instanceTypes, remaining)
-			if len(instanceTypes) == 0 {
-				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed limits for nodepool: %q", nodeClaimTemplate.NodePoolName))
-				continue
-			} else if len(nodeClaimTemplate.InstanceTypeOptions) != len(instanceTypes) {
-				log.FromContext(ctx).V(1).WithValues("NodePool", klog.KRef("", nodeClaimTemplate.NodePoolName)).Info(fmt.Sprintf("%d out of %d instance types were excluded because they would breach limits",
-					len(nodeClaimTemplate.InstanceTypeOptions)-len(instanceTypes), len(nodeClaimTemplate.InstanceTypeOptions)))
+		nodeClaim, err := NewNodeClaimForPod(
+			ctx,
+			nodeClaimTemplate,
+			s.topology,
+			s.daemonOverhead[nodeClaimTemplate],
+			s.remainingResources[nodeClaimTemplate.NodePoolName],
+			pod,
+			s.cachedPodData[pod.UID],
+			s.reservedOfferingMode,
+		)
+		if err != nil {
+			if nodeClaim != nil {
```

Comment:
This strikes me as a weird semantic and I think also leaks hostname values in our topology code just glancing at it -- because we return nil in some scenarios where we have already registered the topology

Reply:
Yeah, you're definitely right about the issue with `Destroy`. The refactor I did here shouldn't be necessary anymore with the current way reservations are tracked, so I'm going to revert this set of changes.